test/test_grid.py

```diff
@@ -778,3 +778,72 @@ def test_build_face_edges_connectivity_fillvalues(self):
         self.assertTrue(
             np.array_equal(res_face_nodes_connectivity,
                            uds._ds["Mesh2_face_nodes"].values))
+
+    def test_node_face_connectivity_from_verts(self):
+        """Test generating Grid.Mesh2_node_faces from array input."""
+        node_latlon_degree = [[162., 30], [216., 30], [70., 30], [162., -30],
+                              [216., -30], [70., -30]]
+
+        face_nodes_conn_index = np.array([[3, 4, 5, ux.INT_FILL_VALUE],
+                                          [3, 0, 2, 5], [3, 4, 1, 0],
+                                          [0, 1, 2, ux.INT_FILL_VALUE]])
+        face_nodes_conn_lonlat = np.full(
+            (face_nodes_conn_index.shape[0], face_nodes_conn_index.shape[1], 2),
+            ux.INT_FILL_VALUE)
+
+        for i, face_nodes_conn_index_row in enumerate(face_nodes_conn_index):
+            for j, node_index in enumerate(face_nodes_conn_index_row):
+                if node_index != ux.INT_FILL_VALUE:
+                    face_nodes_conn_lonlat[i,
+                                           j] = node_latlon_degree[node_index]
+
+        vgrid = ux.Grid(face_nodes_conn_lonlat,
+                        vertices=True,
+                        islatlon=True,
+                        concave=False)
+        expected = np.array([
+            np.array([0, 1, ux.INT_FILL_VALUE]),
+            np.array([1, 3, ux.INT_FILL_VALUE]),
+            np.array([0, 1, 2]),
+            np.array([1, 2, 3]),
+            np.array([0, 2, ux.INT_FILL_VALUE]),
+            np.array([2, 3, ux.INT_FILL_VALUE])
+        ])
```

Comment:
Two things this `expected`: 

1. These magic numbers cannot be understood by the reader of the code. Maybe a geometric justification or something else as a proof that this is the expected node-face connectivity is needed
2. I may be wrong, but my calculation on paper gives me the following node-face connectivity instead (this also shows the need for the ask (1) ):
```
[[1,2,3],
  [2,3],
  [1,3],
  [0,1,2],
  [0,2],
  [0,1]]
```

which is the same with your expected result but has different node order

Reply:
The input will be re-ordered inside the `Grid.build_face_nodes_connectivity` while the topology is the same. We when checking with the expectation, we need to make sure the `Grid._node_face_connectivity` coheres with the `Grid._face_nodes_connectivity`. But I will try to add more explanation in the codes

| Input |         |   | Actually Store in `Grid._face_nodes_connectivity` |         |
|-------|---------|---|---------------------------------------------------|---------|
| Face  | Node    |   | Face                                              | Node    |
| 0     | 3,4,5   |   | 0                                                 | 2,4,0   |
| 1     | 3,0,2,5 |   | 1                                                 | 2,3,1,0 |
| 2     | 3,4,1,0 |   | 2                                                 | 2,4,5,3 |
| 3     | 0,1,2   |   | 3                                                 | 3,5,1   |


| Expected `Grid._node_face_connectivity` |       |
|-----------------------------------------|-------|
| Node                                    | Face  |
| 0                                       | 0, 1  |
| 1                                       | 1,3   |
| 2                                       | 0,1,2 |
| 3                                       | 1,2,3 |
| 4                                       | 0,2   |
| 5                                       | 2,3   |
|                                         |       |

Reply:
Thanks! There is no `Grid.build_face_nodes_connectivity` but I got you meant `Grid.Mesh2_face_nodes`.